lib/reply.js

```diff
@@ -138,9 +138,30 @@ Reply.prototype.send = function (payload) {
 
   if (this[kReplySerializer] !== null) {
     payload = this[kReplySerializer](payload)
-  } else if (hasContentType === false || contentType.indexOf('application/json') > -1) {
-    if (hasContentType === false || contentType.indexOf('charset') === -1) {
+
+  // The indexOf below also matches custom json mimetypes such as 'application/hal+json' or 'application/ld+json'
+  } else if (hasContentType === false || contentType.indexOf('json') > -1) {
+    if (hasContentType === false) {
```

Comment:
How difficult would it be to reorganize branching logic to only check ` if (hasContentType === false)` once?

Reply:
I will have to duplicate [lines 166-169](https://github.com/fastify/fastify/pull/2309/files#diff-d2fdf63c1d7f1d9f68270b6dfa8c930cR166-R169), but we can have only one checking. What do you think about it?

Reply:
Not sure. On a second thought, double check was already there to begin with, so probably we can just leave it as-is.

Reply:
Ops, my mistake. Just a minute.

Reply:
@kibertoad I do not like that duplication, but a possible solution:

```javascript
  } else if (hasContentType === false) {
    this[kReplyHeaders]['content-type'] = CONTENT_TYPE.JSON

    if (typeof payload !== 'string') {
      preserializeHook(this, payload)
      return this
    }
  } else if (contentType.indexOf('json') > -1) {
    // If hasContentType === true, we have a JSON mimetype
    if (contentType.indexOf('charset') === -1) {
      // If we have simply application/json instead of a custom json mimetype
      if (contentType.indexOf('/json') > -1) {
        this[kReplyHeaders]['content-type'] = CONTENT_TYPE.JSON
      } else {
        const currContentType = this[kReplyHeaders]['content-type']
        // We extract the custom mimetype part (e.g. 'hal+' from 'application/hal+json')
        const customJsonType = currContentType.substring(
          currContentType.indexOf('/'),
          currContentType.indexOf('json') + 4
        )

        // We ensure we set the header to the proper JSON content-type if necessary
        // (e.g. 'application/hal+json' instead of 'application/json')
        this[kReplyHeaders]['content-type'] = CONTENT_TYPE.JSON.replace('/json', customJsonType)
      }
    }
    if (typeof payload !== 'string') {
      preserializeHook(this, payload)
      return this
    }
  }
```

Reply:
@kibertoad I will leave it as is then. :D

Reply:
Sorry for the late reply. Either of two options look fine to me.

Reply:
Ok. If you or other reviewers prefer later I can update it before the merge. :D